File: androidp1/app/src/main/java/com/preappointment1/app/data/DeviceAuth.kt
package com.preappointment1.app.data

import android.content.Context
import android.util.Base64
import android.util.Log
import com.preappointment1.app.data.api.ApiClient
import com.preappointment1.app.data.model.DeviceChallengeRequest
import com.preappointment1.app.data.model.DeviceRegisterRequest
import com.preappointment1.app.data.model.DeviceVerifyRequest
import org.bouncycastle.crypto.generators.Ed25519KeyPairGenerator
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.bouncycastle.crypto.KeyGenerationParameters
import retrofit2.HttpException
import java.security.SecureRandom

object DeviceAuth {
    private const val TAG = "LPM_DEVICE_AUTH"
    private const val PREFS = "lpm_device_keys"

    private fun privateKeyKey(deviceId: String): String {
        val safe = deviceId.replace(Regex("[^a-zA-Z0-9._-]"), "_")
        return "private_key_$safe"
    }

    private fun bytesToB64Url(bytes: ByteArray): String {
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }

    private data class KeyPairMaterial(val publicKey: String, val privateKey: ByteArray)

    private fun getOrCreateKeyPair(context: Context, deviceId: String): KeyPairMaterial {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val stored = prefs.getString(privateKeyKey(deviceId), null)
        if (stored != null) {
            val priv = Base64.decode(stored, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
            val pub = Ed25519PrivateKeyParameters(priv, 0).generatePublicKey().encoded
            return KeyPairMaterial(bytesToB64Url(pub), priv)
        }
        val generator = Ed25519KeyPairGenerator()
        generator.init(KeyGenerationParameters(SecureRandom(), 255))
        val kp = generator.generateKeyPair()
        val priv = (kp.private as Ed25519PrivateKeyParameters).encoded
        val pub = (kp.public as Ed25519PublicKeyParameters).encoded
        prefs.edit().putString(privateKeyKey(deviceId), bytesToB64Url(priv)).apply()
        return KeyPairMaterial(bytesToB64Url(pub), priv)
    }

    private fun sign(privateKey: ByteArray, message: String): String {
        val signer = Ed25519Signer()
        signer.init(true, Ed25519PrivateKeyParameters(privateKey, 0))
        val bytes = message.toByteArray(Charsets.UTF_8)
        signer.update(bytes, 0, bytes.size)
        return bytesToB64Url(signer.generateSignature())
    }

    private fun challengePayload(deviceId: String, nonce: String) = "challenge:$deviceId:$nonce"

    private fun registerPayload(deviceId: String, publicKey: String, timestamp: Long, nonce: String) =
        "register:$deviceId:$publicKey:$timestamp:$nonce"

    private fun recoverPayload(deviceId: String, publicKey: String, timestamp: Long, nonce: String) =
        "recover:$deviceId:$publicKey:$timestamp:$nonce"

    suspend fun signIn(context: Context): Boolean {
        val binding = DeviceIdentity.hardwareBindingId(context)
        val deviceId = binding
        val platform = DeviceIdentity.platform()
        val keys = getOrCreateKeyPair(context, deviceId)

        return try {
            verifyExisting(deviceId, binding, keys)
        } catch (e: HttpException) {
            when (e.code()) {
                404 -> tryRegister(deviceId, binding, platform, keys)
                else -> {
                    Log.e(TAG, "Verify failed (${e.code()})", e)
                    false
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Sign in failed", e)
            false
        }
    }

    private suspend fun verifyExisting(deviceId: String, binding: String, keys: KeyPairMaterial): Boolean {
        val challenge = try {
            ApiClient.authApiService.deviceChallenge(
                DeviceChallengeRequest(deviceId = deviceId, intent = "verify")
            )
        } catch (e: HttpException) {
            if (e.code() == 404) throw e
            throw e
        }

        val signature = sign(keys.privateKey, challengePayload(deviceId, challenge.nonce))
        return try {
            val response = ApiClient.authApiService.deviceVerify(
                DeviceVerifyRequest(deviceId = deviceId, nonce = challenge.nonce, signature = signature)
            )
            SessionManager.saveTokens(response.accessToken, response.refreshToken)
            Log.d(TAG, "Device verified")
            true
        } catch (e: HttpException) {
            if (e.code() == 401) {
                tryRecover(deviceId, binding, keys)
            } else {
                throw e
            }
        }
    }

    private suspend fun tryRegister(
        deviceId: String,
        binding: String,
        platform: String,
        keys: KeyPairMaterial
    ): Boolean {
        val challenge = ApiClient.authApiService.deviceChallenge(
            DeviceChallengeRequest(deviceId = deviceId, intent = "register")
        )
        val timestamp = System.currentTimeMillis()
        val signature = sign(
            keys.privateKey,
            registerPayload(deviceId, keys.publicKey, timestamp, challenge.nonce)
        )
        val response = ApiClient.authApiService.deviceRegister(
            DeviceRegisterRequest(
                deviceId = deviceId,
                publicKey = keys.publicKey,
                timestamp = timestamp,
                nonce = challenge.nonce,
                signature = signature,
                hardwareBindingId = binding,
                hardwarePlatform = platform
            )
        )
        SessionManager.saveTokens(response.accessToken, response.refreshToken)
        Log.d(TAG, "Device registered")
        return true
    }

    private suspend fun tryRecover(
        deviceId: String,
        binding: String,
        keys: KeyPairMaterial
    ): Boolean {
        val challenge = ApiClient.authApiService.deviceChallenge(
            DeviceChallengeRequest(deviceId = deviceId, intent = "recover")
        )
        val timestamp = System.currentTimeMillis()
        val signature = sign(
            keys.privateKey,
            recoverPayload(deviceId, keys.publicKey, timestamp, challenge.nonce)
        )
        val response = ApiClient.authApiService.deviceRecover(
            DeviceRegisterRequest(
                deviceId = deviceId,
                publicKey = keys.publicKey,
                timestamp = timestamp,
                nonce = challenge.nonce,
                signature = signature,
                hardwareBindingId = binding,
                hardwarePlatform = DeviceIdentity.platform()
            )
        )
        SessionManager.saveTokens(response.accessToken, response.refreshToken)
        Log.d(TAG, "Device recovered")
        return true
    }
}

File: iosp1/P1/P1/Network/DeviceAuth.swift
import Foundation
import CryptoKit

enum DeviceAuth {
    private static let prefsSuite = "lpm_device_keys"

    private struct KeyPairMaterial {
        let publicKey: String
        let privateKey: Curve25519.Signing.PrivateKey
    }

    private static func privateKeyKey(_ deviceId: String) -> String {
        let safe = deviceId.replacingOccurrences(of: "[^a-zA-Z0-9._-]", with: "_", options: .regularExpression)
        return "private_key_\(safe)"
    }

    private static func bytesToB64Url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func getOrCreateKeyPair(deviceId: String) throws -> KeyPairMaterial {
        let defaults = UserDefaults.standard
        if let stored = defaults.string(forKey: privateKeyKey(deviceId)),
           let privData = Data(base64Encoded: stored.base64URLToBase64()),
           let privateKey = try? Curve25519.Signing.PrivateKey(rawRepresentation: privData) {
            let pub = privateKey.publicKey.rawRepresentation
            return KeyPairMaterial(publicKey: bytesToB64Url(pub), privateKey: privateKey)
        }
        let privateKey = Curve25519.Signing.PrivateKey()
        defaults.set(bytesToB64Url(privateKey.rawRepresentation), forKey: privateKeyKey(deviceId))
        let pub = privateKey.publicKey.rawRepresentation
        return KeyPairMaterial(publicKey: bytesToB64Url(pub), privateKey: privateKey)
    }

    private static func sign(_ privateKey: Curve25519.Signing.PrivateKey, message: String) throws -> String {
        let sig = try privateKey.signature(for: Data(message.utf8))
        return bytesToB64Url(sig)
    }

    static func signIn() async -> Bool {
        let binding = DeviceIdentity.hardwareBindingId()
        let deviceId = binding
        do {
            let keys = try getOrCreateKeyPair(deviceId: deviceId)
            return try await verifyExisting(deviceId: deviceId, binding: binding, keys: keys)
        } catch let error as ApiError {
            if case .serverError(404) = error {
                do {
                    let keys = try getOrCreateKeyPair(deviceId: deviceId)
                    return try await register(deviceId: deviceId, binding: binding, keys: keys)
                } catch {
                    print("LPM_DEVICE_AUTH: register failed \(error)")
                    return false
                }
            }
            print("LPM_DEVICE_AUTH: signIn failed \(error)")
            return false
        } catch {
            print("LPM_DEVICE_AUTH: signIn failed \(error)")
            return false
        }
    }

    private static func verifyExisting(deviceId: String, binding: String, keys: KeyPairMaterial) async throws -> Bool {
        let challenge = try await ApiService.shared.deviceChallenge(
            request: DeviceChallengeRequest(deviceId: deviceId, intent: "verify")
        )
        let signature = try sign(keys.privateKey, message: challengePayload(deviceId: deviceId, nonce: challenge.nonce))
        do {
            let response = try await ApiService.shared.deviceVerify(
                request: DeviceVerifyRequest(deviceId: deviceId, nonce: challenge.nonce, signature: signature)
            )
            SessionManager.shared.saveTokens(access: response.accessToken, refresh: response.refreshToken)
            print("LPM_DEVICE_AUTH: verified")
            return true
        } catch let error as ApiError {
            if case .serverError(401) = error {
                return try await recover(deviceId: deviceId, binding: binding, keys: keys)
            }
            throw error
        }
    }

    private static func register(deviceId: String, binding: String, keys: KeyPairMaterial) async throws -> Bool {
        let challenge = try await ApiService.shared.deviceChallenge(
            request: DeviceChallengeRequest(deviceId: deviceId, intent: "register")
        )
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        let signature = try sign(
            keys.privateKey,
            message: registerPayload(deviceId: deviceId, publicKey: keys.publicKey, timestamp: timestamp, nonce: challenge.nonce)
        )
        let response = try await ApiService.shared.deviceRegister(
            request: DeviceRegisterRequest(
                deviceId: deviceId,
                publicKey: keys.publicKey,
                timestamp: timestamp,
                nonce: challenge.nonce,
                signature: signature,
                hardwareBindingId: binding,
                hardwarePlatform: DeviceIdentity.platform
            )
        )
        SessionManager.shared.saveTokens(access: response.accessToken, refresh: response.refreshToken)
        print("LPM_DEVICE_AUTH: registered")
        return true
    }

    private static func recover(deviceId: String, binding: String, keys: KeyPairMaterial) async throws -> Bool {
        let challenge = try await ApiService.shared.deviceChallenge(
            request: DeviceChallengeRequest(deviceId: deviceId, intent: "recover")
        )
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        let signature = try sign(
            keys.privateKey,
            message: recoverPayload(deviceId: deviceId, publicKey: keys.publicKey, timestamp: timestamp, nonce: challenge.nonce)
        )
        let response = try await ApiService.shared.deviceRecover(
            request: DeviceRegisterRequest(
                deviceId: deviceId,
                publicKey: keys.publicKey,
                timestamp: timestamp,
                nonce: challenge.nonce,
                signature: signature,
                hardwareBindingId: binding,
                hardwarePlatform: DeviceIdentity.platform
            )
        )
        SessionManager.shared.saveTokens(access: response.accessToken, refresh: response.refreshToken)
        print("LPM_DEVICE_AUTH: recovered")
        return true
    }

    private static func challengePayload(deviceId: String, nonce: String) -> String {
        "challenge:\(deviceId):\(nonce)"
    }

    private static func registerPayload(deviceId: String, publicKey: String, timestamp: Int64, nonce: String) -> String {
        "register:\(deviceId):\(publicKey):\(timestamp):\(nonce)"
    }

    private static func recoverPayload(deviceId: String, publicKey: String, timestamp: Int64, nonce: String) -> String {
        "recover:\(deviceId):\(publicKey):\(timestamp):\(nonce)"
    }
}

private extension String {
    func base64URLToBase64() -> String {
        var base = replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let pad = 4 - base.count % 4
        if pad < 4 { base += String(repeating: "=", count: pad) }
        return base
    }
}
